File: project/6.EX_総合理解度チェック/java/java_ex6_ex_4/src/java_ex6_ex_4/Ex6_ex_4.java
package java_ex6_ex_4;

abstract class Widget {
	int weight = 0;
	String color;
}

class WidgetA extends Widget {
	WidgetA() {
		this.weight = 4; 
		this.color = "Red";
	}
}

class WidgetB extends Widget {
	String color = "Blue";
	WidgetB() {
		this.weight = 1;
		super.color = "Pink";
	}
}

class WidgetC extends Widget {
	WidgetC() {
		super.weight = 5; 
		this.color = "Yellow";
	}
}

class WidgetD extends Widget {
	WidgetD() {
		super.weight = 17; 
		super.color = "Green";
	}
}

public class Ex6_ex_4 {

	public static void main(String[] args) {
		Widget[] ws = { new WidgetA(),
					    new WidgetB(),		// color = Pink
					    new WidgetC(),
					    new WidgetD(),
					    new WidgetA(),
					    new WidgetB()};		// color = Pink
		
		// wsインスタンスでは、スーパークラスの変数colorが参照されている
		// ↓のws2インスタンスでは、サブクラスの変数colorが参照されている
		WidgetB ws2 = new WidgetB();		// color = Blue
		
		int sum = 0;
		for( Widget w : ws ) {
			System.out.println( w.color );
			sum += w.weight;
		}
		
		
		System.out.println( ws2.color );
		
		System.out.println( sum );
		
	}

}

File: project/6.EX_総合理解度チェック/py/py_ex6_ex_4/src/ex6_ex_4.py
'''
Created on 2023/05/14

@author: Keroichi.T
'''
from abc import abstractmethod, ABCMeta

class Widget( metaclass = ABCMeta ):
    weight = 10
    color = ""
    
    @abstractmethod
    def __init__(self):
        pass
    
class WidgetA( Widget ):    
    def __init__(self):
        self.weight = 4
        self.color ="Red"
        
class WidgetB( Widget ): 
    color = "Blue"

    def __init__(self):
        self.weight = 1
        self.color ="Pink"
        
class WidgetC( Widget ):    
    def __init__(self):
        self.weight = 5
        self.color ="Yellow"
        
class WidgetD( Widget ):    
    def __init__(self):
        self.weight = 17
        self.color ="Green"
        
def main():
    ws = [ WidgetA(),
           WidgetB(),
           WidgetC(),
           WidgetD(),
           WidgetA(),
           WidgetB() ]

    smy = 0
    for w in ws:
        print( w.color )
        smy += w.weight
    
    print( smy )
    print( sum( [ w.weight for w in ws ] ) )
    

if __name__ == '__main__':
    main()